drgn_tools/meminfo.py

```diff
@@ -0,0 +1,678 @@
+    return mm_consts
+
+
+def print_val_kb(prog: Program, text: str, num: int) -> None:
+    """
+    Produce the formatted output that matches the output in /proc/meminfo.
+
+    :param prog: drgn program
+    :param text: The output item's name, e.g, "MemTotal"
+    :param num: The output item's value
+    """
+    if num < 0:
+        return
+
+    num_kb = num << (get_mm_constants(prog)["PAGE_SHIFT"] - 10)
+    num_kb_str = str(num_kb)
+    print(f"{text + ':': <15} {num_kb_str: >8} kB")
+
+
+def get_active_numa_nodes(prog: Program) -> List[Object]:
+    """
+    Get a list of all active NUMA nodes.
+
+    :returns: a list of ``struct pglist_data *`` objects.
+    """
+    active_numa_nodes = []
+    node_data = prog["node_data"]
+    for i in range(get_mm_constants(prog)["MAX_NUM_NODES"]):
+        if node_data[i].value_() != 0x0:
+            active_numa_nodes.append(node_data[i])
+    return active_numa_nodes
+
+
+def for_each_node_zone(prog: Program, node: Object) -> Iterator[Object]:
+    """
+    Iterate over all memory zones in a NUMA node.
+
+    :param prog: drgn program
+    :param node: The drgn object of a NUMA node's ``struct pglist_data``.
+    :returns: Iterator of ``struct zone *`` objects.
+    """
+    node_zones = node.member_("node_zones")
+    max_num_zones = prog["__MAX_NR_ZONES"].value_()
+    for j in range(max_num_zones):
+        if node_zones[j].value_() != 0x0:
+            yield node_zones[j]
+
+
+def for_each_zone(prog: Program) -> Iterator[Object]:
+    """
+    Iterate over all zones in a system that contains 1+ NUMA nodes.
+
+    :returns: Iterator of ``struct zone *`` objects.
+    """
+    for node in get_active_numa_nodes(prog):
+        for zone in for_each_node_zone(prog, node):
+            yield zone
+
+
+def for_each_hstate(prog: Program) -> Iterator[Object]:
+    """
+    Iterate over all hugepage pools in a system.
+    Note: each |hstate| represents a pool for a certain unit size hugepages of a NUMA node.
+
+    :returns: Iterator of ``struct hstate`` objects.
+    """
+    hstates = prog["hstates"].read_()
+    for i in range(prog["hugetlb_max_hstate"].value_()):
+        yield hstates[i]
+
+
+def get_total_available_pages(prog: Program) -> int:
+    """
+    Get an estimation of the amount of memory available for future use.
+    This includes the number of free memory pages and the number of pages
+    that can be reclaimed from caches.
+
+    :returns: The number of available memory in pages.
+    """
+    global_stats = get_global_mm_stats(prog)
+
+    total_free_pages = global_stats["NR_FREE_PAGES"]
+    total_reserve_pages = prog["totalreserve_pages"].value_()
+    available_pages = total_free_pages - total_reserve_pages
+
+    # Calculate the global low watermark.
+    low_wmark = 0
+    for zone in for_each_zone(prog):
+        if has_member(zone, "_watermark"):
+            zone_low_wmark = zone.member_("_watermark")[
+                prog.constant("WMARK_LOW")
+            ]
```

Comment:
nit: (which applies throughout the PR) - the `member_()` method is almost always unnecessary. You could just write `zone._watermark[prog.constant("WMARK_LOW")]` here. The [`member_()`](https://drgn.readthedocs.io/en/latest/api_reference.html#drgn.Object.member_) method exists mainly to aid you in instances where a method on `Object` has the same name as a struct field. For example, if there was a field named `value_` in a struct, then you'd need to use `obj.member_("value_")` since `value_` is already a method of drgn's Object.

That said, there's nothing technically wrong with using it! I guess I'm just pointing it out to let you know, and it is slightly easier to read without it.

Reply:
Done!